### PR TITLE
Adds a mechanism in vSphere Cloud Provider to get credentials from Kubernetes secrets

### PR DESCRIPTION
**Pull request by abrarshivani**
Currently, vCenter credentials are stored in plain text in vsphere.conf. This PR adds a mechanism in vSphere Cloud Provider to get vCenter credentials from Kubernetes secrets.
**Workflow:**
1. Create vsphere.conf file with ```secret-name``` and ```secret-namespace```.
	```
	[Global]
	insecure-flag = 1
	secret-name = "vcconf"
	secret-namespace = "kube-system"

	[VirtualCenter "10.160.45.119"]
	port = 443
	datacenters = k8s-dc-1

	[Workspace]
	server = 10.160.45.119
	datacenter = k8s-dc-1
	default-datastore = sharedVMFS-0
	folder = Discovered virtual machine
	```
2. Launch Kubernetes cluster with vSphere Cloud Provider Configured.
3. Create secret with vCenter credentials.
	a. Create base64 encoding for username and password:
	username:
	```	
		> echo -n 'admin' | base64
		YWRtaW4= 
	```
	password:
	```
		> echo -n 'vsphere' | base64
		dnNwaGVyZQ==
	```

	b. kubectl create -f vccredentials.yaml
	```
		#vccredentials.yaml
		apiVersion: v1
		kind: Secret
		metadata:
			name: vcconf
		type: Opaque
		data:
			10.192.44.199.username: YWRtaW4=
			10.192.44.199.password: dnNwaGVyZQ==
	```
**Note:**
Secrets info can be provided with both(old and new) vSphere Cloud provider configuration formats.

**Tests Done:**
- [x] vSphere Cloud Provider unit test.
- [x] Volume lifecyle with Username and Password in vsphere.conf (for backward compability)
- [x] Volume lifecyle with secrets information in vsphere.conf.
- [x] Update secrets

### CODE REVIEW

**Comment by tpepper** [comment]
Are you planning to squash the very long series of commits, refractors, fixes and refixes into a more concise series of commits which cleanly and incrementally introduces the new code?

**Comment by tpepper** [comment]
So really I'm looking for more than a simplistic squash.  Your diffstat is:

+923 −69 

That's more code than typically should be in a single commit, in order to make for simpler review and bisection and future potential problem determination.

If you squash the final state of your work, please then decompose the squash into a short series of patches with introduce the new functionality as a logic progression.

**Comment by abrarshivani** [comment]
@tpepper Done

File: pkg/cloudprovider/providers/vsphere/nodemanager.go [hunk 1]
```diff
@@ -370,3 +360,70 @@ func (nodeInfo *NodeInfo) VM() *vclib.VirtualMachine {
 	}
 	return nodeInfo.vm
 }
+
+func (nm *NodeManager) vcConnect(ctx context.Context, vsphereInstance *VSphereInstance) error {
+	err := vsphereInstance.conn.Connect(ctx)
+	credentialManager := nm.CredentialManager()
+	if err == nil || !vclib.IsInvalidCredentialsError(err) || credentialManager == nil {
+		return err
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/nodemanager.go]: Lets log the err before return.

File: pkg/cloudprovider/providers/vsphere/nodemanager.go [hunk 2]
```diff
@@ -370,3 +360,70 @@ func (nodeInfo *NodeInfo) VM() *vclib.VirtualMachine {
...
+	}
+	glog.V(4).Infof("Invalid credentials. Cannot connect to server %q. " +
+		"Fetching credentials from secrets.", vsphereInstance.conn.Hostname)
+
+	// Get latest credentials from SecretCredentialManager
+	credentials, err := credentialManager.GetCredential(vsphereInstance.conn.Hostname)
+	if err != nil {
+		return err
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/nodemanager.go]: log err

File: pkg/cloudprovider/providers/vsphere/nodemanager.go (hunk 1 above)
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/nodemanager.go]: log err

File: pkg/cloudprovider/providers/vsphere/nodemanager.go [hunk 3]
```diff
@@ -370,3 +360,70 @@ func (nodeInfo *NodeInfo) VM() *vclib.VirtualMachine {
...
+	}
+	vsphereInstance.conn.UpdateCredentials(credentials.User, credentials.Password)
+	return vsphereInstance.conn.Connect(ctx)
+}
+
+// GetNodeInfo returns a NodeInfo which datacenter, vm and vc server ip address.
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/nodemanager.go]: GetNodeInfo => GetNodeInfoWithNodeObject

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 4]
```diff
@@ -305,31 +356,48 @@ func populateVsphereInstanceMap(cfg *VSphereConfig) (map[string]*VSphereInstance
...
+			if !isSecretInfoProvided {
+				if vcConfig.User == "" {
+					vcConfig.User = cfg.Global.User
+				}
+				if vcConfig.Password == "" {
+					vcConfig.Password = cfg.Global.Password
+				}
+				if vcConfig.User == "" {
+					msg := fmt.Sprintf("vcConfig.User is empty for vc %s!", vcServer)
+					glog.Error(msg)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: lets use errorf

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 5]
```diff
@@ -305,31 +356,48 @@ func populateVsphereInstanceMap(cfg *VSphereConfig) (map[string]*VSphereInstance
...
-				glog.Error(msg)
-				return nil, errors.New(msg)
-			}
-			if vcConfig.Password == "" {
-				msg := fmt.Sprintf("vcConfig.Password is empty for vc %s!", vcServer)
-				glog.Error(msg)
-				return nil, errors.New(msg)
+
+			if !isSecretInfoProvided {
+				if vcConfig.User == "" {
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: check for cfg.Global.User is not empty in this block. same for cfg.Global.Password.

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 6]
```diff
@@ -305,31 +356,48 @@ func populateVsphereInstanceMap(cfg *VSphereConfig) (map[string]*VSphereInstance
...
+				}
+				if vcConfig.Password == "" {
+					msg := fmt.Sprintf("vcConfig.Password is empty for vc %s!", vcServer)
+					glog.Error(msg)
+					return nil, ErrPasswordMissing
+				}
+			} else {
+				if vcConfig.User != "" {
+					msg := fmt.Sprintf("vcConfig.User for server %s and Secret info provided. VCP will use secret to get credentials", vcServer)
+					glog.Warning(msg)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Warning -> Warningf

File: pkg/cloudprovider/providers/vsphere/credentialmanager_test.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+				err := secretInformer.Informer().GetIndexer().Add(expected.secret)
+				if err != nil {
+					t.Fatalf("Failed to add secret to internal cache: %v", err)
+				}
+			case getCredentialsOp:
+				expected := test.expectedValues[ntest].(GetCredentialsTest)
+				credential, err := secretCredentialManager.GetCredential(expected.server)
+				t.Logf("Retrieving credentials for server %s", expected.server)
+				if err != expected.err {
+					t.Fatalf("Fail to get credentials with error: %s", err)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/credentialmanager_test.go]: For error lets use `%v`

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 5 above)
Comment (abrarshivani) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: Done

File: pkg/cloudprovider/providers/vsphere/credentialmanager.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	return *credential, found
+}
+
+func (cache *SecretCache) parseSecret() error {
+	cache.cacheLock.Lock()
+	defer cache.cacheLock.Unlock()
+	return parseConfig(cache.Secret.Data, cache.VirtualCenter)
+}
+
+func parseConfig(data map[string][]byte, config map[string]*Credential) error {
```
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/credentialmanager.go]: Can you a doc on what is the expected config format

File: pkg/cloudprovider/providers/vsphere/credentialmanager.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+		statusErr, ok := err.(*apierrors.StatusError)
+		if (ok && statusErr.ErrStatus.Code != http.StatusNotFound) || !ok {
+			return nil, err
+		}
+		// Handle secrets deletion
+		glog.Warningf("secret %q not found in namespace %q", secretCredentialManager.SecretName, secretCredentialManager.SecretNamespace)
+	}
+	// Cases:
+	// 1. Secret Deleted finding credentials from cache
+	// 2. Secret Not Added at a first place will return error
```
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/credentialmanager.go]: Not sure what the first point is. Irrespective of the secret deleted, you get the credentials from Cache. Is it right?
Rewrite point 2. as "Return error if Secret is not added" and 3. "Return error is the secret doesn't contain any credentials".

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 10]
```diff
@@ -247,19 +277,40 @@ func newWorkerNode() (*VSphere, error) {
...
 	if cfg.VirtualCenter == nil || len(cfg.VirtualCenter) == 0 {
 		glog.V(4).Infof("Config is not per virtual center and is in old format.")
-		if cfg.Global.User == "" {
-			glog.Error("Global.User is empty!")
-			return nil, errors.New("Global.User is empty!")
-		}
-		if cfg.Global.Password == "" {
-			glog.Error("Global.Password is empty!")
-			return nil, errors.New("Global.Password is empty!")
+		if !isSecretInfoProvided {
```
[the remark this answers is not in the document]
Comment (abrarshivani) [reply pkg/cloudprovider/providers/vsphere/vsphere.go]: Added note before initializing vSphere structure.

**Pull request closed without merging**